Mathlib/Algebra/Category/ModuleCat/Colimits.lean

```diff
@@ -0,0 +1,388 @@
+#align Module.colimits.cocone_naturality ModuleCat.Colimits.cocone_naturality
+
+@[simp]
+theorem cocone_naturality_components (j j' : J) (f : j ⟶ j') (x : F.obj j) :
+    (coconeMorphism F j') (F.map f x) = (coconeMorphism F j) x := by
+  rw [← cocone_naturality F f]
+  rfl
+#align Module.colimits.cocone_naturality_components ModuleCat.Colimits.cocone_naturality_components
+
+/-- The cocone over the proposed colimit module. -/
+def colimitCocone : Cocone F where
+  pt := colimit F
+  ι := { app := coconeMorphism F }
+#align Module.colimits.colimit_cocone ModuleCat.Colimits.colimitCocone
+
+/-- The function from the free module on the diagram to the cone point of any other cocone. -/
+@[simp]
+def descFunLift (s : Cocone F) : Prequotient F → s.pt
+  | Prequotient.of j x => (s.ι.app j) x
+  | zero => 0
+  | neg x => -descFunLift _ x
+  | add x y => descFunLift _ x + descFunLift _ y
+  | smul s x => s • descFunLift _ x
+#align Module.colimits.desc_fun_lift ModuleCat.Colimits.descFunLift
+
+/-- The function from the colimit module to the cone point of any other cocone. -/
+def descFun (s : Cocone F) : ColimitType F → s.pt := by
+  fapply Quot.lift
+  · exact descFunLift F s
+  · intro x y r
+    induction' r with h₁ r_x r_y r_h r_ih r_x r_y r_z r_h r_k r_ih_h r_ih_k r_j r_j' r_f r_x j j x
+      j x y j s x u v r r_ih u v w r r_ih u v w r r_ih s u v r r_ih <;> try dsimp
+    -- refl
+    -- · rfl -- porting note: `dsimp` (above) now closes this
+    -- symm
+    · exact r_ih.symm
+    -- trans
+    · exact Eq.trans r_ih_h r_ih_k
+    -- map
+    · exact s.w_apply r_f r_x -- porting note: `simp` failed
+    -- zero
+    · simp
+    -- neg
+    · simp
+    -- add
+    · simp
+    -- smul,
+    · simp
+    -- neg_1
+    · rw [r_ih]
+    -- add_1
+    · rw [r_ih]
+    -- add_2
+    · rw [r_ih]
+    -- smul_1
+    · rw [r_ih]
+    -- zero_add
+    · rw [zero_add]
+    -- add_zero
+    · rw [add_zero]
+    -- add_left_neg
+    · rw [add_left_neg]
+    -- add_comm
+    · rw [add_comm]
+    -- add_assoc
+    · rw [add_assoc]
+    -- one_smul
+    · rw [one_smul]
+    -- mul_smul
+    · rw [mul_smul]
+    -- smul_add
+    · rw [smul_add]
+    -- smul_zero
+    · rw [smul_zero]
+    -- add_smul
+    · rw [add_smul]
+    -- zero_smul
+    · rw [zero_smul]
+#align Module.colimits.desc_fun ModuleCat.Colimits.descFun
+
+/-- The group homomorphism from the colimit module to the cone point of any other cocone. -/
+def descMorphism (s : Cocone F) : colimit F ⟶ s.pt where
+  toFun := descFun F s
+  map_smul' s x := by rcases x <;> rfl
+  map_add' x y := by rcases x <;> rcases y <;> rfl
+#align Module.colimits.desc_morphism ModuleCat.Colimits.descMorphism
+
+/-- Evidence that the proposed colimit is the colimit. -/
+def colimitCoconeIsColimit : IsColimit (colimitCocone F) where
+  fac := sorry -- Workaround https://github.com/JLimperg/aesop/issues/57 Remove after `uniq` done.
+  desc s := descMorphism F s
+  uniq s m w := by
+    ext ⟨⟨x_j, x_x⟩⟩
```

Comment:
I now realise `ext` is not powerful enough and we really need `induction`, as in the [Mathlib3 proof](https://github.com/leanprover-community/mathlib/blob/9d013ad8430ddddd350cff5c3db830278ded3c79/src/algebra/category/Module/colimits.lean#L354).

However I cannot get `induction` or `induction'` to fire in Mathlib4. Furthermore, although it is possible to replace the Mathlib3 use of `induction x` with `refine quot.rec (λ x_1, _) _ x, clear x, rename x_1 x,` successfully, I find `refine' Quot.rec (fun x_1 => ?_) ?_ x` fails in Mathlib4.

I'm going to leave this PR until I have learned Lean 4 properly (or until someone else shows me what to do here).